File: exercises/go/user_repository.go
package main

import (
	"database/sql"
	"fmt"
	"time"

	_ "github.com/lib/pq"
)

// UserRepository handles database operations for User entities
type UserRepository struct {
	db *sql.DB
}

// NewUserRepository creates a new UserRepository instance
func NewUserRepository(db *sql.DB) *UserRepository {
	return &UserRepository{db: db}
}

// CreateUser creates a new user in the database
func (r *UserRepository) CreateUser(user *User) (*User, error) {
	query := `
		INSERT INTO users (username, email, first_name, last_name, created_at, updated_at)
		VALUES ($1, $2, $3, $4, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP)
		RETURNING id, created_at, updated_at`

	var id int
	var createdAt, updatedAt time.Time

	err := r.db.QueryRow(query, user.Username, user.Email, user.FirstName, user.LastName).
		Scan(&id, &createdAt, &updatedAt)
	if err != nil {
		return nil, fmt.Errorf("failed to create user: %w", err)
	}

	user.ID = id
	user.CreatedAt = createdAt
	user.UpdatedAt = updatedAt

	return user, nil
}

// FindByID finds a user by ID
func (r *UserRepository) FindByID(id int) (*User, error) {
	query := `
		SELECT id, username, email, first_name, last_name, created_at, updated_at
		FROM users WHERE id = $1`

	user := &User{}
	err := r.db.QueryRow(query, id).Scan(
		&user.ID, &user.Username, &user.Email, &user.FirstName, &user.LastName,
		&user.CreatedAt, &user.UpdatedAt,
	)

	if err != nil {
		if err == sql.ErrNoRows {
			return nil, nil // User not found
		}
		return nil, fmt.Errorf("failed to find user by ID: %w", err)
	}

	return user, nil
}

// FindByUsername finds a user by username
func (r *UserRepository) FindByUsername(username string) (*User, error) {
	query := `
		SELECT id, username, email, first_name, last_name, created_at, updated_at
		FROM users WHERE username = $1`

	user := &User{}
	err := r.db.QueryRow(query, username).Scan(
		&user.ID, &user.Username, &user.Email, &user.FirstName, &user.LastName,
		&user.CreatedAt, &user.UpdatedAt,
	)

	if err != nil {
		if err == sql.ErrNoRows {
			return nil, nil // User not found
		}
		return nil, fmt.Errorf("failed to find user by username: %w", err)
	}

	return user, nil
}

// FindByEmail finds a user by email
func (r *UserRepository) FindByEmail(email string) (*User, error) {
	query := `
		SELECT id, username, email, first_name, last_name, created_at, updated_at
		FROM users WHERE email = $1`

	user := &User{}
	err := r.db.QueryRow(query, email).Scan(
		&user.ID, &user.Username, &user.Email, &user.FirstName, &user.LastName,
		&user.CreatedAt, &user.UpdatedAt,
	)

	if err != nil {
		if err == sql.ErrNoRows {
			return nil, nil // User not found
		}
		return nil, fmt.Errorf("failed to find user by email: %w", err)
	}

	return user, nil
}

// UpdateUser updates an existing user
func (r *UserRepository) UpdateUser(user *User) error {
	query := `
		UPDATE users 
		SET username = $1, email = $2, first_name = $3, last_name = $4, updated_at = CURRENT_TIMESTAMP
		WHERE id = $5`

	result, err := r.db.Exec(query, user.Username, user.Email, user.FirstName, user.LastName, user.ID)
	if err != nil {
		return fmt.Errorf("failed to update user: %w", err)
	}

	rowsAffected, err := result.RowsAffected()
	if err != nil {
		return fmt.Errorf("failed to get rows affected: %w", err)
	}

	if rowsAffected == 0 {
		return fmt.Errorf("user with ID %d not found", user.ID)
	}

	// Update the UpdatedAt timestamp
	err = r.db.QueryRow("SELECT updated_at FROM users WHERE id = $1", user.ID).Scan(&user.UpdatedAt)
	if err != nil {
		return fmt.Errorf("failed to get updated timestamp: %w", err)
	}

	return nil
}

// DeleteUser deletes a user by ID
func (r *UserRepository) DeleteUser(id int) error {
	query := "DELETE FROM users WHERE id = $1"

	result, err := r.db.Exec(query, id)
	if err != nil {
		return fmt.Errorf("failed to delete user: %w", err)
	}

	rowsAffected, err := result.RowsAffected()
	if err != nil {
		return fmt.Errorf("failed to get rows affected: %w", err)
	}

	if rowsAffected == 0 {
		return fmt.Errorf("user with ID %d not found", id)
	}

	return nil
}

// FindAllUsers retrieves all users
func (r *UserRepository) FindAllUsers() ([]*User, error) {
	query := `
		SELECT id, username, email, first_name, last_name, created_at, updated_at
		FROM users ORDER BY id`

	rows, err := r.db.Query(query)
	if err != nil {
		return nil, fmt.Errorf("failed to query users: %w", err)
	}
	defer rows.Close()

	var users []*User
	for rows.Next() {
		user := &User{}
		err := rows.Scan(
			&user.ID, &user.Username, &user.Email, &user.FirstName, &user.LastName,
			&user.CreatedAt, &user.UpdatedAt,
		)
		if err != nil {
			return nil, fmt.Errorf("failed to scan user: %w", err)
		}
		users = append(users, user)
	}

	if err = rows.Err(); err != nil {
		return nil, fmt.Errorf("error iterating over rows: %w", err)
	}

	return users, nil
}

// CountUsers returns the total number of users
func (r *UserRepository) CountUsers() (int, error) {
	query := "SELECT COUNT(*) FROM users"

	var count int
	err := r.db.QueryRow(query).Scan(&count)
	if err != nil {
		return 0, fmt.Errorf("failed to count users: %w", err)
	}

	return count, nil
}

// DeleteAllUsers deletes all users (useful for test cleanup)
func (r *UserRepository) DeleteAllUsers() error {
	query := "DELETE FROM users"

	_, err := r.db.Exec(query)
	if err != nil {
		return fmt.Errorf("failed to delete all users: %w", err)
	}

	return nil
}

// Close closes the database connection
func (r *UserRepository) Close() error {
	if r.db != nil {
		return r.db.Close()
	}
	return nil
}

File: exercises/nodejs/src/user-repository.ts
import { Pool, PoolClient } from 'pg';
import { User, UserEntity } from './user';

export class UserRepository {
  private pool: Pool;

  constructor(pool: Pool) {
    this.pool = pool;
  }

  /**
   * Creates a new user in the database
   */
  async createUser(user: UserEntity): Promise<UserEntity> {
    const query = `
      INSERT INTO users (username, email, first_name, last_name, created_at, updated_at)
      VALUES ($1, $2, $3, $4, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP)
      RETURNING id, created_at, updated_at`;

    const values = [user.username, user.email, user.firstName || null, user.lastName || null];

    const client = await this.pool.connect();
    try {
      const result = await client.query(query, values);
      const row = result.rows[0];

      user.id = row.id;
      user.createdAt = row.created_at;
      user.updatedAt = row.updated_at;

      return user;
    } finally {
      client.release();
    }
  }

  /**
   * Finds a user by ID
   */
  async findById(id: number): Promise<UserEntity | null> {
    const query = `
      SELECT id, username, email, first_name, last_name, created_at, updated_at
      FROM users WHERE id = $1`;

    const client = await this.pool.connect();
    try {
      const result = await client.query(query, [id]);
      
      if (result.rows.length === 0) {
        return null;
      }

      return this.mapRowToUser(result.rows[0]);
    } finally {
      client.release();
    }
  }

  /**
   * Finds a user by username
   */
  async findByUsername(username: string): Promise<UserEntity | null> {
    const query = `
      SELECT id, username, email, first_name, last_name, created_at, updated_at
      FROM users WHERE username = $1`;

    const client = await this.pool.connect();
    try {
      const result = await client.query(query, [username]);
      
      if (result.rows.length === 0) {
        return null;
      }

      return this.mapRowToUser(result.rows[0]);
    } finally {
      client.release();
    }
  }

  /**
   * Finds a user by email
   */
  async findByEmail(email: string): Promise<UserEntity | null> {
    const query = `
      SELECT id, username, email, first_name, last_name, created_at, updated_at
      FROM users WHERE email = $1`;

    const client = await this.pool.connect();
    try {
      const result = await client.query(query, [email]);
      
      if (result.rows.length === 0) {
        return null;
      }

      return this.mapRowToUser(result.rows[0]);
    } finally {
      client.release();
    }
  }

  /**
   * Updates an existing user
   */
  async updateUser(user: UserEntity): Promise<UserEntity> {
    const query = `
      UPDATE users 
      SET username = $1, email = $2, first_name = $3, last_name = $4, updated_at = CURRENT_TIMESTAMP
      WHERE id = $5
      RETURNING updated_at`;

    const values = [
      user.username,
      user.email,
      user.firstName || null,
      user.lastName || null,
      user.id,
    ];

    const client = await this.pool.connect();
    try {
      const result = await client.query(query, values);
      
      if (result.rows.length === 0) {
        throw new Error(`User with ID ${user.id} not found`);
      }

      user.updatedAt = result.rows[0].updated_at;
      return user;
    } finally {
      client.release();
    }
  }

  /**
   * Deletes a user by ID
   */
  async deleteUser(id: number): Promise<boolean> {
    const query = 'DELETE FROM users WHERE id = $1';

    const client = await this.pool.connect();
    try {
      const result = await client.query(query, [id]);
      return result.rowCount > 0;
    } finally {
      client.release();
    }
  }

  /**
   * Retrieves all users
   */
  async findAllUsers(): Promise<UserEntity[]> {
    const query = `
      SELECT id, username, email, first_name, last_name, created_at, updated_at
      FROM users ORDER BY id`;

    const client = await this.pool.connect();
    try {
      const result = await client.query(query);
      return result.rows.map(row => this.mapRowToUser(row));
    } finally {
      client.release();
    }
  }

  /**
   * Counts the total number of users
   */
  async countUsers(): Promise<number> {
    const query = 'SELECT COUNT(*) as count FROM users';

    const client = await this.pool.connect();
    try {
      const result = await client.query(query);
      return parseInt(result.rows[0].count, 10);
    } finally {
      client.release();
    }
  }

  /**
   * Deletes all users (useful for test cleanup)
   */
  async deleteAllUsers(): Promise<void> {
    const query = 'DELETE FROM users';

    const client = await this.pool.connect();
    try {
      await client.query(query);
    } finally {
      client.release();
    }
  }

  /**
   * Executes a transaction
   */
  async executeTransaction<T>(callback: (client: PoolClient) => Promise<T>): Promise<T> {
    const client = await this.pool.connect();
    try {
      await client.query('BEGIN');
      const result = await callback(client);
      await client.query('COMMIT');
      return result;
    } catch (error) {
      await client.query('ROLLBACK');
      throw error;
    } finally {
      client.release();
    }
  }

  /**
   * Maps a database row to a UserEntity
   */
  private mapRowToUser(row: any): UserEntity {
    const user = new UserEntity();
    user.id = row.id;
    user.username = row.username;
    user.email = row.email;
    user.firstName = row.first_name;
    user.lastName = row.last_name;
    user.createdAt = row.created_at;
    user.updatedAt = row.updated_at;
    return user;
  }

  /**
   * Closes the connection pool
   */
  async close(): Promise<void> {
    await this.pool.end();
  }
}
